File: src/GameCore/Social/FactionSystem.cs
namespace GameCore.Social;

/// <summary>阵营枚举：游骑兵/冒险公会/商盟/暗影/圣殿/平民。</summary>
public enum FactionId
{
    Ranger,         // 游骑兵：狩猎/森林/驯兽任务
    Adventurer,     // 冒险者公会：讨伐/护送/探索
    Merchant,       // 商盟：运输/矿石/制造
    Shadow,         // 暗影：潜行/暗杀/盗窃
    Holy,           // 圣殿：治疗/驱魔/誓约
    Citizen,        // 平民声望：救助/说服/日常
}

/// <summary>阵营声望等级。</summary>
public enum RepLevel { Hostile = -1, Neutral = 0, Friendly = 1, Honored = 2, Revered = 3 }

/// <summary>阵营声望管理。单例在 GameWorld.Bootstrap 中初始化。</summary>
public sealed class FactionSystem
{
    public static FactionSystem Instance { get; private set; }

    private readonly Dictionary<FactionId, int> _rep = new();

    public FactionSystem()
    {
        Instance = this;
        foreach (FactionId f in Enum.GetValues<FactionId>())
            _rep[f] = 0;   // 初始中立
    }

    /// <summary>获取声望值。</summary>
    public int Get(FactionId fid) => _rep.GetValueOrDefault(fid, 0);

    /// <summary>从 reward 键 "reputation_xxx" 中加声望（xxx=ranger/adventurer/merchant/shadow/holy/citizen）。</summary>
    public void Add(string rewardKey, int amount)
    {
        var fid = ParseFaction(rewardKey);
        if (fid == null) return;
        Add(fid.Value, amount);
    }

    public void Add(FactionId fid, int amount)
    {
        _rep[fid] = Math.Clamp(_rep.GetValueOrDefault(fid) + amount, -999, 999);
    }

    /// <summary>获取声望等级。</summary>
    public RepLevel GetLevel(FactionId fid)
    {
        var v = Get(fid);
        if (v < -50) return RepLevel.Hostile;
        if (v < 100) return RepLevel.Neutral;
        if (v < 300) return RepLevel.Friendly;
        if (v < 600) return RepLevel.Honored;
        return RepLevel.Revered;
    }

    /// <summary>获取对应[声望Lv]标签ID（用于 VM/Condition 检查）。</summary>
    public string GetTagId(FactionId fid) => GetLevel(fid) switch
    {
        RepLevel.Hostile => $"{fid}声望敌视",
        RepLevel.Friendly => $"{fid}声望友善",
        RepLevel.Honored => $"{fid}声望尊敬",
        RepLevel.Revered => $"{fid}声望崇敬",
        _ => $"{fid}声望中立"
    };

    /// <summary>初始化/读档恢复。</summary>
    public void Restore(Dictionary<FactionId, int> saved)
    {
        foreach (var (k, v) in saved)
            _rep[k] = v;
    }

    public Dictionary<FactionId, int> Snapshot() => new(_rep);

    private static FactionId? ParseFaction(string key)
    {
        if (string.IsNullOrEmpty(key)) return null;
        if (key.StartsWith("reputation_")) key = key["reputation_".Length..];
        return key.ToLowerInvariant() switch
        {
            "ranger" => FactionId.Ranger,
            "adventurer" => FactionId.Adventurer,
            "merchant" => FactionId.Merchant,
            "shadow" => FactionId.Shadow,
            "holy" => FactionId.Holy,
            "citizen" => FactionId.Citizen,
            _ => null
        };
    }
}

File: canglan-backend/canglan-world/src/main/java/com/canglan/world/social/FactionSystem.java
package com.canglan.world.social;

import java.util.EnumMap;
import java.util.Locale;
import java.util.Map;

/**
 * FactionSystem — 阵营声望管理。对应 C# FactionSystem
 * （静态 Instance 改为实例持有；ReputationChanged 事件改为监听器回调）。
 */
public final class FactionSystem {

    /** 声望变化监听：(阵营, 本次变化量, 变化后声望等级)。 */
    @FunctionalInterface
    public interface ReputationListener {
        void onChanged(FactionId faction, int delta, RepLevel newLevel);
    }

    private final Map<FactionId, Integer> rep = new EnumMap<>(FactionId.class);
    private ReputationListener listener;

    public FactionSystem() {
        for (FactionId f : FactionId.values()) rep.put(f, 0);   // 初始中立
    }

    public void setListener(ReputationListener listener) { this.listener = listener; }

    /** 获取声望值。 */
    public int get(FactionId fid) { return rep.getOrDefault(fid, 0); }

    /** 从 reward 键 "reputation_xxx" 中加声望（xxx=ranger/adventurer/merchant/shadow/holy/citizen）。 */
    public void add(String rewardKey, int amount) {
        FactionId fid = parseFaction(rewardKey);
        if (fid == null) return;
        add(fid, amount);
    }

    public void add(FactionId fid, int amount) {
        rep.put(fid, Math.max(-999, Math.min(999, get(fid) + amount)));
        if (listener != null) listener.onChanged(fid, amount, getLevel(fid));
    }

    /** 获取声望等级。 */
    public RepLevel getLevel(FactionId fid) {
        int v = get(fid);
        if (v < -50) return RepLevel.HOSTILE;
        if (v < 100) return RepLevel.NEUTRAL;
        if (v < 300) return RepLevel.FRIENDLY;
        if (v < 600) return RepLevel.HONORED;
        return RepLevel.REVERED;
    }

    /** 获取对应[声望Lv]标签ID（用于 VM/Condition 检查）。 */
    public String getTagId(FactionId fid) {
        return switch (getLevel(fid)) {
            case HOSTILE -> fid.displayName() + "声望敌视";
            case FRIENDLY -> fid.displayName() + "声望友善";
            case HONORED -> fid.displayName() + "声望尊敬";
            case REVERED -> fid.displayName() + "声望崇敬";
            default -> fid.displayName() + "声望中立";
        };
    }

    /** 初始化/读档恢复。 */
    public void restore(Map<FactionId, Integer> saved) {
        if (saved == null) return;
        for (Map.Entry<FactionId, Integer> kv : saved.entrySet()) rep.put(kv.getKey(), kv.getValue());
    }

    public Map<FactionId, Integer> snapshot() { return new EnumMap<>(rep); }

    private static FactionId parseFaction(String key) {
        if (key == null || key.isEmpty()) return null;
        if (key.startsWith("reputation_")) key = key.substring("reputation_".length());
        return switch (key.toLowerCase(Locale.ROOT)) {
            case "ranger" -> FactionId.RANGER;
            case "adventurer" -> FactionId.ADVENTURER;
            case "merchant" -> FactionId.MERCHANT;
            case "shadow" -> FactionId.SHADOW;
            case "holy" -> FactionId.HOLY;
            case "citizen" -> FactionId.CITIZEN;
            default -> null;
        };
    }
}
